Feature/zlib support

jeff-kilbride opened the pull request:
This PR adds support for gzip values. I have tried to keep it as simple as possible.

I have added a `gzip` key to the various `options` configuration objects. Setting this key to `true` enables a default gzip implementation using lowest compression / best speed settings. Setting this key to an object simply passes that object to the `gzip`/`gunzip` methods as configuration. I am not currently doing any validation of these settings, if an object is passed. This should only be used for advanced configuration of the underlying `gzip` methods. More validation could be added, but I would like some input on this first given the extra code required. Obviously, passing `false` or omitting the `gzip` key disables this feature.

The changes I have made are working for me locally with my own tests and with a previous fork of the project from about a week ago. However, I am having issues running `npm test` with the latest fork from today. I will gladly add some Jasmine tests, when this is resolved. I wanted to go ahead and create a PR, so you could see the changes I have made.

mrister commented on index.js:
```diff
@@ -153,21 +185,28 @@ function redisStore(args) {
+    }
+    options.parse = true;
+
+    var gzip = (options.gzip || options.gzip === false) ? options.gzip : redisOptions.gzip;
```
Would  more simple comparison work the same? like:
 `var gzip = options.gzip ? options.gzip : redisOptions.gzip;`

jeff-kilbride replied:
Unfortunately, no, if you want users to be able to turn gzip on/off at the set/get level. If `redisOptions.gzip` is truthy (object) and the user wants to turn it off for a particular set/get, right now they can pass `false`:

`myCache.set('myCacheKey', 'myCacheValue', { gzip: false }, function...`
`myCache.get('myCacheKey', { gzip: false }, function...`

The more simple comparison would not catch that. The `options.gzip` would evaluate to `false` and the `redisOptions.gzip` would be used. That's why I have an explicit test for `false`.  It's the same type of test that you are doing in the `set` method for the `options.ttl` value, in order to catch zero:

`var ttl = (options.ttl || options.ttl === 0) ? options.ttl : redisOptions.ttl;`

That's actually where I got the idea.   :)

mrister replied:
ok :-)

mrister commented on index.js:
```diff
@@ -12,6 +13,10 @@ var redisUrl = require('redis-url');
  * @param {Number} args.port - The Redis server port
  * @param {Number} args.db - The Redis server db
  * @param {function} args.isCacheableValue - function to override built-in isCacheableValue function (optional)
+ * @param {boolean|Object} args.gzip - (optional) Boolean / Config Object for gzip / gunzip compression.
```
It would be good to have this documentation with examples in readme

jeff-kilbride replied:
Ok, I can take a look at that.

jeff-kilbride commented:
I'm a little confused by the `codacy/pr` checks. It's complaining about things like `strings must be in double-quotes`. Are you moving in that direction? Should I change my code to pass the `codacy` checks?

PuKoren commented:
@jeff-kilbride don't worry about codacy, it even have different policies that ours (and it is yelling because of issues you didn't even added). We will remove it in near future in favor of sonar

jeff-kilbride commented:
I have merged the latest changes and added a few fixes. I still need to add gzip specific tests and something for the Readme. However, if you set `gzip: true` in the options for the current tests in this branch, all tests are now passing. 

I will try to get this done in the next few days, but I have some other pressing things I need to work on.

mrister commented:
@jeff-kilbride Thank you very much for your contributions, anytime is good as long as you can find it to finish this feature.

jeff-kilbride commented:
Sorry for the delay! I have added gzip specific tests and updated the Readme file. 

I wrapped all the existing tests in a new `describe` statement and put the gzip tests in their own `describe` to keep them separate. So, the `redis-store-spec.js` file shows a lot of changes, but they are nearly all whitespace. I copied all the existing `get/set` tests into `redis-store-zlib-spec.js` and added a few new ones. Two of my new tests grab the string length of the compressed values and print out the compression percentage to the console. I thought this output was interesting, but it can easily be removed if you prefer not to have it.

Also, while editing the Readme file, I noticed a possible error. In your examples, you are passing a `ttl` option:
```js
var ttl = 5;

redisCache.set('foo', 'bar', ttl, function(err) {
...
```
The `ttl` being passed is a `Number`, but the `set` method is expecting an `Object`. I believe this should be:
```js
var ttl = { ttl: 5 };

redisCache.set('foo', 'bar', ttl, function(err) {
...
```
or
```js
var ttl = 5;

redisCache.set('foo', 'bar', { ttl: ttl }, function(err) {
...
```
If this is correct, I can make these changes to the Readme and push another commit. Let me know.

PuKoren commented:
Thansk for your contribution @jeff-kilbride !

About the error in the readme, you can do it on a separate PR if you want to, this one will be focused on your gzip feature

Thanks for the refactoring of tests and for the feature! We will review it asap to plan a publication

PuKoren commented on index.js:
```diff
@@ -190,16 +228,38 @@ function redisStore(args) {
+          conn.setex(key, ttl, val, handleResponse(conn, cb));
+        }
+        else {
+          conn.set(key, val, handleResponse(conn, cb));
```
l.249 l.252 l.258 and l.261 are duplicated, maybe we could refactor that in a function or something? What do you think?

jeff-kilbride replied:
It could be refactored. I did that on purpose to clearly separate the new code from the existing. I wanted you to be able to see my changes easily.

To refactor, the `conn.set` and `conn.setex` would need to be moved into a new function. What do you think we should name that function?

PuKoren replied:
We could name it `set` as of new versions of redis, it seems that `setex` will be deprecated as `set` is used with same parameters: http://redis.io/commands/set

So i would say, `set`, and add a layer for previous versions compatibility inside it

What do you think guys? @dial-once/developers

jeff-kilbride replied:
Yes, I agree. The `setex` should be replaced, but I'm not sure if the `node-redis` module supports the new format for the `set` command, yet. Has anyone tested that?

PuKoren replied:
Not on my side. But we can keep the same code inside our `set` function anyway, I was just referencing this to chose the correct naming for the function

jeff-kilbride replied:
Ok. Makes sense. I'm still a little confused, though. Sorry! 

Right now, there is `self.set()`. In order to refactor the duplicate code in lines `248-262`, there will need to be a new function that basically does this:
```js
function newSet(key, value, ttl, cb) {
  if (typeof ttl === 'function') {
    cb = ttl;
    ttl = '';
  }

  if (ttl) {
    conn.setex(key, ttl, value, handleResponse(conn, cb));
  }
  else {
    conn.set(key, value, handleResponse(conn, cb));
  }
}
```
Then lines `248-253` and `257-262` would each be replaced by:
```js
  newSet(key, gzVal, ttl, cb);
// or
  newSet(key, val, ttl, cb);
```
Is this what you are thinking of? If so, what would you like to call `newSet`? It's basically a helper function for `self.set`.

PuKoren replied:
I was thinking of something like this directly inside the set:

```js
      function persist(err, value) {
        if (gzErr) {
          return cb && cb(gzErr);
        }
        if (ttl) {
            conn.setex(key, ttl, value, handleResponse(conn, cb));
          }
          else {
            conn.set(key, value, handleResponse(conn, cb));
          }
      }

      if (gzip) {
        zlib.gzip(val, gzip, persist);
      }
      else {
        persist(null, val);
      }
```

Not sure if it is best practice, but I find it more readable than having multiple set commands inside the main object

Wooh, this project needs a good cleanup :sun_with_face:

jeff-kilbride replied:
Ah... ok. Makes sense. Pushed a new change.

mrister commented on test/lib/redis-store-spec.js:
```diff
@@ -6,460 +6,466 @@ var assert = require('assert');
       host: config.redis.host,
       port: config.redis.port,
-      password: config.redis.auth_pass,
+      auth_pass: config.redis.auth_pass,
```
`auth_pass` is being easily deprecated we should promote usage of `password`. https://github.com/NodeRedis/node_redis#options-object-properties

PuKoren replied:
That is right, we also have some tests about it. A little refactor would be good, I'll add a ticket do to it in a separate PR

mrister replied:
ok

mrister commented on test/lib/redis-store-zlib-spec.js:
```diff
@@ -0,0 +1,372 @@
+      auth_pass: config.redis.auth_pass,
+      db: config.redis.db,
+      ttl: config.redis.ttl,
+      gzip: true
```
Now thinking about it, I wanted to ask what you guys @PuKoren @jeff-kilbride would think about making this a little bit more abstract. To be able to easily plug in other compressors/decompressors and gzip being the first choice for now.

So instead of having gzip:true having a `compression: 'gzip'`(or default to it if compression: true). Based on the value of compression property user can pick which one they need (maybe in future there would be others) or do their own.  I know that this means a little big refactor, but it should open up options. (maybe compression is a bad name, but we can adapt and chose another)

PuKoren replied:
That is a good idea, will avoid having to deal with a lot of options later on and open the road to the compression pipelining :smile:

jeff-kilbride replied:
It's an interesting idea. It would would need to be modularized and the correct module would need to be lazy-loaded based on the options provided. It could wind up being a good-sized refactor, though.

You may want to find out how much interest there is in the community for other compression algorithms, first. I can think of a few -- snappy, LZ4, LZO, etc... -- but I don't know how many people understand the differences between them or would want to use them. The nice thing about `gzip` is that it's built into the Node `Zlib` module and has no further dependencies. It's not the fastest algorithm, but it's a decent tradeoff, especially when set to it's lowest compression / fastest speed configuration -- which is what I typically use.

Anyway, I am open to the idea. I always prefer my tools being more flexible, rather than less.  :)

jeff-kilbride commented:
This latest commit refactors the duplicate code in the `set` method, as discussed above.

PuKoren commented:
Anyone have some time to refactor compression options [as discussed here](https://github.com/dial-once/node-cache-manager-redis/pull/44#pullrequestreview-1109890)?

If not, then it LGTM for next feature release

mrister commented:
@PuKoren As I suggested the refactor, maybe I should do it :-) I will. Just hope there is no big hurry on this.

mrister commented:
compression pipelining.. interesting :)

jeff-kilbride commented:
I will continue to help whenever I can. Maybe it would be a good idea to start a new issue for the refactor.

@mrister  I would enjoy hearing your ideas on how this should be implemented.

mrister commented:
@jeff-kilbride Jes perhaps another issue. Well my initial Idea was to have a list of codecs (compressors and decomposers) in which you can optionally add or remove a codec as a dependency. The gzip one will serve as a reference implementation of that. (So a separate module that will get loaded when needed by this libs user)
After that community can create other that we can ship with this module or just have their own. The list would allow for pipelining if so required.

jeff-kilbride commented:
@mrister For pipelining, you mean being able to call multiple compression algorithms for the same value?

mrister commented:
@jeff-kilbride Yes basically the output of one would be input for the other algorithm. I am not sure if this is a real feature request or @PuKoren just joking :-) Either way it can be done.

PuKoren commented:
@mrister was not joking but maybe it is a bit early, because:
  - it may be totally useless (CPU cost/speed/compression ratio ridiculous)
  - it can be added later quite easily without breaking change (convert compression property to an array)

I think that if you have some time to do the generic compression stuff, you should not worry about pipelining as it may take too long for what it's worth

mrister commented:
OK then, still I think We need to do this before the release (Separate PR and issue) to not introduce breaking changes.

jeff-kilbride commented:
Hi All,

I have merged the latest changes to `develop` into this branch and updated some of my tests to get them working again.

Is there any update to the progress on adding additional compression implementations? It should be fairly easy to refactor this branch to avoid breaking changes in the future. The `args.gzip` variable could be renamed to `args.compression` (or something similar, if you have another idea). It could still accept either a `boolean` or `object`. A value of `true` could default to the `gzip` implementation. An `object` could have a `type` property with the name of the compression implementation -- `gzip` or any future additions -- as well as implementation specific parameters to be passed to the underlying compression layer (the same way `gzip` parameters are currently being passed).

These changes would be cosmetic, for now, because `gzip` would remain the only compression algorithm currently supported. However, it would then be possible to add additional compression implementations without breaking the interface in future releases. It would require a bigger rewrite to handle pluggable compression modules, but that could be handled by anyone desiring to use something other than `gzip`.

If this sounds like the correct approach, I can make these cosmetic changes and commit them in this PR.

PuKoren commented:
I am for the cosmetic change

What about you guys @dial-once/developers ?

jkernech commented:
I'm for it too 👍

mrister commented:
yup go for it @jeff-kilbride 👍

jeff-kilbride commented:
I have changed the `opts.gzip` parameter to `opts.compress` with the following default values:

```js
{
  type: 'gzip',
  params: {
    level: zlib.Z_BEST_SPEED
  }
}
```

The idea is that, in the future as pluggable modules are added, the `type` property will be the name of the module / implementation and the (optional) `params` property will contain any module specific values to be passed to the underlying compression algorithm. Obviously, right now, the `type` property is ignored, since only `gzip` is supported. However, the code can now be refactored to support pluggable modules without having to change the interface.

Let me know what you guys think.

PuKoren commented:
Great work @jeff-kilbride, its very good to have implemented it with options/type. We can do anything in the future with this implementation 👍

mrister commented on test/lib/redis-store-zlib-spec.js:
```diff
@@ -0,0 +1,380 @@
+      compress: true,
+      isCacheableValue: function (val) {
+        // allow undefined
+        if (val === undefined)
```
minor for code style we should use parenthesis on if else.

mrister commented:
All good @jeff-kilbride nice work! minor code style changes.

jeff-kilbride commented:
@mrister  Requested code style changes fixed.

mrister commented on index.js:
```diff
@@ -179,16 +221,36 @@ function redisStore(args) {
+      if (compress) {
+        zlib.gzip(val, compress.params || {}, persist);
+      }
+      else {
```
same here

jeff-kilbride commented:
@mrister  Done.
